File: TrabajoAcademicoNet/SoftProgPersistencia/DAOImpl/DAOImplPersona/ClienteDAOImpl.cs
using MySqlX.XDevAPI;
using PUCP.Inf30.SoftProg.Modelo.PersonaModelo;
using PUCP.Inf30.SoftProg.Modelo.PersonaModelo.PersonaUtils;
using PUCP.Inf30.SoftProg.Persistencia.DAO.DAOPersona;
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.Common;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace PUCP.Inf30.SoftProg.Persistencia.DAOImpl.DAOImplPersona
{
    public class ClienteDAOImpl : BaseDAOImpl<Cliente>, IClienteDAO
    {
        

        protected override DbCommand ComandoActualizar(DbConnection conn, Cliente cliente)
        {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "actualizarCliente";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_idCliente", DbType.Int32, cliente.Id);
            this.AgregarParametroEntrada(cmd, "@p_lineaCredito", DbType.Double, cliente.LineaCredito);
            this.AgregarParametroEntrada(cmd, "@p_Categoria", DbType.String, cliente.Categoria);
            this.AgregarParametroEntrada(cmd, "@p_numeroDePedido_Historico", DbType.Int32, cliente.NumeroPedidosHistorico);
            this.AgregarParametroEntrada(cmd, "@p_numeroDePedido_MensualPro", DbType.Int32, cliente.NumeroPedidosMensualPro);
            this.AgregarParametroEntrada(cmd, "@p_dni", DbType.String, cliente.Dni);
            this.AgregarParametroEntrada(cmd, "@p_nombre", DbType.String, cliente.Nombre);
            this.AgregarParametroEntrada(cmd, "@p_apellidoPaterno", DbType.String, cliente.ApellidoPaterno);
            this.AgregarParametroEntrada(cmd, "@p_apellidoMaterno", DbType.String, cliente.ApellidoMaterno);
            this.AgregarParametroEntrada(cmd, "@p_genero", DbType.String, cliente.Genero);
            this.AgregarParametroEntrada(cmd, "@p_fechaNacimiento", DbType.DateTime, cliente.FechaNacimiento);
            this.AgregarParametroEntrada(cmd, "@p_telefono", DbType.Int32, cliente.Telefono);
            this.AgregarParametroEntrada(cmd, "@p_Activo", DbType.Boolean, cliente.Activo);

            return cmd;
        }

        protected override DbCommand ComandoCrear(DbConnection conn, Cliente cliente)
        {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "insertarCliente";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_idCliente", DbType.Int32, cliente.Id);
            this.AgregarParametroEntrada(cmd, "@p_lineaCredito", DbType.Double, cliente.LineaCredito);
            this.AgregarParametroEntrada(cmd, "@p_Categoria", DbType.String, cliente.Categoria);
            this.AgregarParametroEntrada(cmd, "@p_numeroDePedido_Historico", DbType.Int32, cliente.NumeroPedidosHistorico);
            this.AgregarParametroEntrada(cmd, "@p_numeroDePedido_MensualPro", DbType.Int32, cliente.NumeroPedidosMensualPro);
            this.AgregarParametroEntrada(cmd, "@p_dni", DbType.String, cliente.Dni);
            this.AgregarParametroEntrada(cmd, "@p_nombre", DbType.String, cliente.Nombre);
            this.AgregarParametroEntrada(cmd, "@p_apellidoPaterno", DbType.String, cliente.ApellidoPaterno);
            this.AgregarParametroEntrada(cmd, "@p_apellidoMaterno", DbType.String, cliente.ApellidoMaterno);
            this.AgregarParametroEntrada(cmd, "@p_genero", DbType.String, cliente.Genero);
            this.AgregarParametroEntrada(cmd, "@p_fechaNacimiento", DbType.DateTime, cliente.FechaNacimiento);
            this.AgregarParametroEntrada(cmd, "@p_telefono", DbType.Int32, cliente.Telefono);
            this.AgregarParametroEntrada(cmd, "@p_Activo", DbType.Boolean, cliente.Activo);

            return cmd;
        }

        protected override DbCommand ComandoEliminar(DbConnection conn, int id)
        {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "eliminarCliente";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_idCliente", DbType.Int32, id);

            return cmd;
        }

        protected override DbCommand ComandoLeer(DbConnection conn, int id)
        {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "buscarCliente";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_idCliente", DbType.Int32, id);

            return cmd;
        }

        protected override DbCommand ComandoLeerTodos(DbConnection conn)
        {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "listarClientes";
            cmd.CommandType = CommandType.StoredProcedure;
            return cmd;
        }

        protected override Cliente MapearModelo(DbDataReader reader)
        {
            return new Cliente
            {
                Id = Convert.ToInt32(reader["id"]),
                Categoria = (CategoriaCliente)Enum.Parse(typeof(CategoriaCliente), Convert.ToString(reader["categoria"])),
                NumeroPedidosHistorico = Convert.ToInt32(reader["numeroPedidosHistorico"]),
                NumeroPedidosMensualPro = Convert.ToInt32(reader["numeroDePedido_MensualPro"]),
                Dni = Convert.ToString(reader["p_dni"]),
                Nombre = Convert.ToString(reader["nombre"]),
                ApellidoPaterno = Convert.ToString(reader["apellidoPaterno"]),
                ApellidoMaterno = Convert.ToString(reader["apellidoMaterno"]),
                Genero = (Genero)Enum.Parse(typeof(Genero), Convert.ToString(reader["genero"])),
                FechaNacimiento = Convert.ToDateTime(reader["fechaNacimiento"]),
                Telefono = Convert.ToInt32(reader["telefono"]),
                Activo = Convert.ToBoolean(reader["activo"]),
                LineaCredito = Convert.ToDouble(reader["lineaCredito"])
            };
        }

        protected DbCommand ComandoBuscarPorDni(DbConnection conn, string dni)
        {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "buscarClientePorDni";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_idCliente", DbType.Int32, dni);

            return cmd;
        }

        public Cliente buscarPorDNI(string dni)
        {
            return EjecutarComando(conn => {
                using (DbCommand cmd = this.ComandoBuscarPorDni(conn, dni))
                using (DbDataReader reader = cmd.ExecuteReader())
                {
                    if (!reader.HasRows)
                    {
                        Console.Error.WriteLine("No se encontro el cliente con dni: " + dni);
                        return default;
                    }

                    reader.Read();
                    return this.MapearModelo(reader);
                }
            });
        }


    }
}

File: Java/TrabajoAcademico/SoftProgDAO/src/main/java/pe/edu/pucp/inf30/softprog/daoimpl/persona/ClienteDAOImpl.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package pe.edu.pucp.inf30.softprog.daoimpl.persona;

import java.sql.CallableStatement;
import java.sql.Connection;
import java.sql.Date;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Types;
import pe.edu.pucp.inf30.softprog.dao.persona.ClienteDAO;
import pe.edu.pucp.inf30.softprog.dao.persona.CuentaUsuarioDAO;
import pe.edu.pucp.inf30.softprog.daoimpl.BaseDAO;
import pe.edu.pucp.inf30.softprog.modelo.persona.Cliente;

/**
 *
 * @author Cristhian Horacio
 */
public class ClienteDAOImpl extends BaseDAO<Cliente> implements ClienteDAO{

    @Override
    protected PreparedStatement comandoCrear(Connection conn, Cliente modelo) throws SQLException {
        String sql = "{call insertarCliente(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        
        cmd.setInt("p_idCliente", modelo.getId());
        cmd.setDouble("p_lineaCredito", modelo.getLineaCredito());
        cmd.setString("p_Categoria", modelo.getCategoriaCliente());
        cmd.setInt("p_numeroDePedido_Historico", modelo.getNumeroPedidosHistorico());
        cmd.setInt("p_numeroDePedido_MensualPro", modelo.getNumeroPedidosMensualPro());
        cmd.setString("p_dni", modelo.getDni());
        cmd.setString("p_nombre", modelo.getNombre());
        cmd.setString("p_apellidoPaterno", modelo.getApellidoPaterno());
        cmd.setString("p_apellidoMaterno", modelo.getApellidoMaterno());
        cmd.setString("p_genero", modelo.getGeneroString());
        cmd.setDate("p_fechaNacimiento", (Date) modelo.getFechaNacimiento());
        cmd.setInt("p_telefono", modelo.getTelefono());
        cmd.setInt("p_Activo", modelo.getActivo());
        cmd.registerOutParameter("p_id", Types.INTEGER);
        
        return cmd;
    }

    @Override
    protected PreparedStatement comandoActualizar(Connection conn, Cliente modelo) throws SQLException {
        String sql = "{call modificarCliente(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        
        cmd.setInt("p_idCliente", modelo.getId());
        cmd.setDouble("p_lineaCredito", modelo.getLineaCredito());
        cmd.setString("p_Categoria", modelo.getCategoriaCliente());
        cmd.setInt("p_numeroDePedido_Historico", modelo.getNumeroPedidosHistorico());
        cmd.setInt("p_numeroDePedido_MensualPro", modelo.getNumeroPedidosMensualPro());
        cmd.setString("p_dni", modelo.getDni());
        cmd.setString("p_nombre", modelo.getNombre());
        cmd.setString("p_apellidoPaterno", modelo.getApellidoPaterno());
        cmd.setString("p_apellidoMaterno", modelo.getApellidoMaterno());
        cmd.setString("p_genero", modelo.getGeneroString());
        cmd.setDate("p_fechaNacimiento", (Date) modelo.getFechaNacimiento());
        cmd.setInt("p_telefono", modelo.getTelefono());
        cmd.setInt("p_Activo", modelo.getActivo());
        
        return cmd;
    }

    @Override
    protected PreparedStatement comandoEliminar(Connection conn, Integer id) throws SQLException {
        String sql = "{call eliminarCliente(?)}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        
        cmd.setInt("p_idCliente", id);
        
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeer(Connection conn, Integer id) throws SQLException {
        String sql = "{call buscarClientePorId(?)}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        
        cmd.setInt("p_idCliente", id);
        
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeerTodos(Connection conn) throws SQLException {
        String sql = "{call listarClientes()}";
        
        CallableStatement cmd = conn.prepareCall(sql);
  
        
        return cmd;
    }

    @Override
    protected Cliente mapearModelo(ResultSet rs) throws SQLException {
        Cliente cliente = new Cliente();
        
        cliente.setId(rs.getInt("idCliente"));
        cliente.setCategoriaCliente(rs.getString("Categoria"));
        cliente.setNumeroPedidosHistorico(rs.getInt("numeroDePedido_Historico"));
        cliente.setNumeroPedidosMensualPro(rs.getInt("numeroDePedido_MensualPro"));
        cliente.setDni(rs.getString("dni"));
        cliente.setNombre(rs.getString("nombre"));
        cliente.setApellidoPaterno(rs.getString("apellidoPaterno"));
        cliente.setApellidoMaterno(rs.getString("apellidoMaterno"));
        cliente.setGeneroString(rs.getString("genero"));
        cliente.setFechaNacimiento(rs.getDate("fechaNacimiento"));
        cliente.setTelefono(rs.getInt("telefono"));
        cliente.setActivoInt(rs.getInt("Activo"));
        cliente.setLineaCredito(rs.getDouble("lineaCredito"));
        
        return cliente;
    }

    protected PreparedStatement comandoBuscarPorDni(Connection conn, String dni) 
            throws SQLException {
        
        String sql = "{call buscarClientePorDni(?)}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setString("p_dni", dni);
        
        return cmd;
    }
    
    @Override
    public Cliente buscarPorDNI(String dni) {
        return ejecutarComando(conn -> {
            try (PreparedStatement cmd = this.comandoBuscarPorDni(conn, dni)){
                ResultSet rs = cmd.executeQuery();
                
                if (!rs.next()) {
                    System.err.println("No se encontro el registro con "
                            + "dni: " + dni);
                    return null;
                }
                
                return this.mapearModelo(rs);
            }           
        });
    }

    public Cliente obtenerPorId(int id) {
        throw new UnsupportedOperationException("Not supported yet."); // Generated from nbfs://nbhost/SystemFileSystem/Templates/Classes/Code/GeneratedMethodBody
    }
    
}
